### ISSUE TITLE
Groupby hash aggregations use sort-based implementation if nested-type columns are used as values

### ISSUE THREAD

**Issue by divyegala** [rapidsai/cudf#14412]
We should be able to use nested-type columns as values and still be able to invoke a hash-based groupby, as hash-based is generally faster so we do not want to be silently using sort-based.

**Comment by bdice** [comment]
This code block is the piece in question:

https://github.com/rapidsai/cudf/blob/b446a6f187241e765c925da1053ece2679313a06/cpp/src/groupby/hash/groupby.cu#L654-L656

@ttnghia, you added this in #13676. Do you know if this fallback is still required, or why?

We discussed this a bit here: https://github.com/rapidsai/cudf/pull/13676#discussion_r1260196703

Side note, I'm punching my past self -- I asked this question on that PR, and never submitted my review: 
![image](https://github.com/rapidsai/cudf/assets/3943761/01e9d6f3-f021-44b0-91cf-235fe57d340e)

**Comment by ttnghia** [comment]
> @ttnghia, you added this in #13676. Do you know if this fallback is still required, or why?

Because hash-based aggregations are implemented for plain type only, using the operator such as `<` instead of user-provided comparator. See `struct update_target_element` in `cpp/include/cudf/detail/aggregation/aggregation.cuh`.

If we want to support hash-based aggregates for nested types, we need to rewrite such `struct update_target_element`.

**Comment by bdice** [comment]
Great, that was helpful. I think we can do this. I think the rough plan would be to preprocess the table so we have device comparators that can be used, pass the preprocessed table info through all the aggregation machinery, and use the device comparator where needed in `update_target_element`. Does that sound right to you?

**Comment by ttnghia** [comment]
Yes that sounds good. Note that we only needs to rework for `ARGMIN` and `ARGMAX`, not for everything thus the amount of work should not be very heavy.

**Comment by PointKernel** [comment]
To clarify, the reason we cannot use hash-based groupby for nested types is that there is currently no way to atomically update nested data on the device due to the lack of direct hardware support for such operations. A possible solution is to use an atomic lock table, which CCCL is expected to support in the future. I will provide links to the relevant GitHub issues or PRs if I come across any.

We should backlog this for now until the atomic lock table becomes available.
